### ISSUE TITLE
Add vehicle position on map

### ISSUE THREAD

**Issue by kalon33** [HSLdevcom/digitransit-ui#461]
It would be an awesome feature to add vehicle position on the map, either real-time position (if available) or simulated one, calculated from theorical timetables (like Transit app does on Android). Apart from a 'gadget' feature, it makes us more patient when waiting for the vehicle, and makes network visualization to another level.

**Comment by tuukka** [comment]
At the moment, we show real-time vehicle positions on the route page (both on the map and in the schematic stop list). We have been discussing this and on the front page may show vehicles only on routes that you have marked as your favourites. On the stop page, we may show all vehicles that are going to stop here. On the itinerary page, we can show the specific vehicles that we instruct you to board.

**Comment by kalon33** [comment]
So if there is no real-time information available, no position is displayed?

All of the things you are discussing are great and relevant in the different use cases I think.

**Comment by tuukka** [comment]
That's correct. For simulation of real-time positions based on static GTFS data, I have been using this: http://14142.net/kartalla/about.html

**Comment by kalon33** [comment]
Is it planned to complete digitransit vehicle position display with something like that, warning the user that this is theorical position?

**Comment by samuliheljo** [comment]
There has been some discussion about showing vehicles' expected location but it's not currently in our roadmap

**Comment by kalon33** [comment]
@samuliheljo is this now on the roadmap? Even if it's not shown on the map (which is IMHO the harder part, as vehicle position has to be extrapolated along the shape between stops, dividing shape length by the number of seconds needed to travel between stops and positioning the point on the shape), indicating vehicle position on the stops view when displaying route based on theorical data (when RT is not available) should be useful to infer bus frequency and where to go to get into it, and be a nice complement to route schedule implemented in #670 ... maybe @apkoponen have ideas for this?

Basically I would imagine 3 possibilities for the vehicle. At stop A when it should be there according to theorical data, at stop B when it should arrive to it (same data source), and in the middle between them after leaving A and before arriving to B, when exact position not available.

Unfortunately, I don't have the knowledge at all to work on that... but I may bounty the whole thing if people wants extra motivation to work on this during their extra time.

**Comment by samuliheljo** [comment]
Situation is same as before. Our primary goal is to move towards real time
information and use that on UI.

Interpolating vehicles where they approximitely are should be possible but
I'm not sure how our product owner feels about that. Real time is really
big thing in Helsinki area during next year and my feeling is that our main
focus is in that.

If there is enough community momentum behind this feature in e.g. August,
we can start discussing about pros and cons and design alternatives for
implementation and see who would be interested in making it reality.

Our availability during summer time is limited as we try to recover
from Beta launch, which by the way happens today!

So, in addition, I want to thank you for all the work you have done with us
:)



On Wednesday, 15 June 2016, Nicolas Derive <notifications@github.com
<javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:

> @samuliheljo <https://github.com/samuliheljo> is this now on the roadmap?
> Even if it's not shown on the map (which is IMHO the harder part, as
> vehicle position has to be extrapolated along the shape between stops,
> dividing shape length by the number of seconds needed to travel between
> stops and positioning the point on the shape), indicating vehicle position
> on the stops view when displaying route based on theorical data (when RT is
> not available) should be useful to infer bus frequency and where to go to
> get into it, and be a nice complement to route schedule implemented in
> #670 <https://github.com/HSLdevcom/digitransit-ui/pull/670> ... maybe
> @apkoponen <https://github.com/apkoponen> have ideas for this?
>
> Basically I would imagine 3 possibilities for the vehicle. At stop A when
> it should be there according to theorical data, at stop B when it should
> arrive to it (same data source), and in the middle between them after
> leaving A and before arriving to B, when exact position not available.
>
> Unfortunately, I don't have the knowledge at all to work on that... but I
> may bounty the whole thing if people wants extra motivation to work on this
> during their extra time.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/HSLdevcom/digitransit-ui/issues/461#issuecomment-226174266>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABPQWgqBRL_sDVIELAKCwR6sLSGKGb5cks5qL_ILgaJpZM4H3zBc>
> .
>


-- 

--
Samuli Heljo
+358 50 3537299

**Comment by kalon33** [comment]
@samuliheljo you're welcome, it's a pleasure to contribute to a project with such a nice community :)

I understand the problem for Helsinki, that's really nice that some cities have interest in opening access to real time data in a usable and standard format. Unfortunately, it's not the case in France, when RT is open, which is rare (not in Paris, in Bordeaux but not linked to GTFS data so unusable in our situation (with OTP notably) and not in a standard format, same thing as Bordeaux for national train service...), mainly because to my mind they want to keep really usable data for themselves to keep the value of it (but that's another subject)... That's why I was talking about alternative solutions, waiting for the situation to improve here... but really, I see that as a fallback when real time data is not available, and it should be displayed as such (maybe using another color or a dedicated message somewhere?)

Thanks for your explanations on this!

Thanks also to all the community and contributors for such a nice work :)

**Comment by apkoponen** [comment]
Technically this would be quite easy to do, but @samuliheljo mentioned, this might be more of a UI / UX issue.

I made pretty straightforward algorithm into [Plunkr](https://plnkr.co/edit/n9G0VvgHjLOQiITwyGnM) that calculates the position of the vehicle based on the schedule or realtime data from this [GraphQL query](http://dev.hsl.fi/graphql/console/?query=%7B%0A%20%20pattern(id%3A%20%22HSL%3A3001R%3A0%3A01%22)%20%7B%0A%20%20%20%20%09geometry%20%7B%0A%20%20%20%20%20%20%20%20lat%0A%20%20%20%20%20%20%20%20lon%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20tripsForDate(serviceDay%3A%20%2220160620%22)%20%7B%0A%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20stoptimes%20%7B%0A%20%20%20%20%20%20%20%20%20%20scheduledArrival%0A%20%20%20%20%20%20%20%20%20%20scheduledDeparture%0A%20%20%20%20%20%20%20%20%20%20realtimeArrival%0A%20%20%20%20%20%20%20%20%20%20realtimeDeparture%0A%20%20%20%20%20%20%20%20%20%20stop%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20lat%0A%20%20%20%20%20%20%20%20%20%20%20%20lon%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%7D%0A%7D).

Logic:

1. Filter those trips that are currently active.
2. For each active trip:
3. Find the segment we are on based on stoptimes (realtime or scheduled).
4. Check if we are staying on a stop, if we are, return the station stop location.
5. If we are in the middle of two stops, find the points closest to our current start and end stops on the geometry line.
6. Get the segment where we are on the geometry line.
7. Calculate our location on this line.

The algorithm returns an array of objects with the following properties:

- location: An object with lat and lon as floats.
- progress: Fragment of how much of the current segment has been travelled as float.
- bearing: The current direction in degrees from north as float.
- moving: Whether the vehicle is moving or in station as boolean.

E.g.
{
"location: {
    "lat": 60.62200750121682,
    "lon": 24.872947068716414
},
"progress": 0.875,
"bearing": -38.58968855167587,
"moving": true

**Comment by samuliheljo** [comment]
Yup. Ux is probably more challenging here, but also probably not undoable.

One option / idea for this is to implement it as a "fake realtime source".

What a mean by that is a completely separate service from UI that would
implement realtime interface but provide data from static timetables. Then
UI would just be configured to use that as a realtime source and magically
it would work and show vehicles in "realtime".

Currently we don't have that clearly defined realtime interface (Like
to say "implement what this document says"), but we would like UI to
support 1..n realtime endpoints where n is as low as possible. We'll see
how this plays out in the future.

So in Helsinki case, we are proceeding towards actual realtime solution,
but some kind of fake realtime service could be interesting contribution
from community.

We haven't discussed about this idea in our team, therefore I shall notify
our people and see what reactions we get.




On Thursday, 16 June 2016, A-P Koponen <notifications@github.com> wrote:

> Technically this would be quite easy to do, but @samuliheljo
> <https://github.com/samuliheljo> mentioned, this might be more of a UI /
> UX issue.
>
> I made pretty straightforward algorithm into Plunkr
> <https://plnkr.co/edit/n9G0VvgHjLOQiITwyGnM> that calculates the position
> of the vehicle based on the schedule or realtime data from this GraphQL
> query
> <http://dev.hsl.fi/graphql/console/?query=%7B%0A%20%20pattern(id%3A%20%22HSL%3A3001R%3A0%3A01%22)%20%7B%0A%20%20%20%20%09geometry%20%7B%0A%20%20%20%20%20%20%20%20lat%0A%20%20%20%20%20%20%20%20lon%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20tripsForDate(serviceDay%3A%20%2220160620%22)%20%7B%0A%20%20%20%20%20%20%20%20id%0A%20%20%20%20%20%20%20%20stoptimes%20%7B%0A%20%20%20%20%20%20%20%20%20%20scheduledArrival%0A%20%20%20%20%20%20%20%20%20%20scheduledDeparture%0A%20%20%20%20%20%20%20%20%20%20realtimeArrival%0A%20%20%20%20%20%20%20%20%20%20realtimeDeparture%0A%20%20%20%20%20%20%20%20%20%20stop%20%7B%0A%20%20%20%20%20%20%20%20%20%20%20%20lat%0A%20%20%20%20%20%20%20%20%20%20%20%20lon%0A%20%20%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20%7D%0A%20%20%7D%0A%7D>
> .
>
> Logic:
>
>    1. Filter those trips that are currently active.
>    2. For each active trip:
>    3. Find the segment we are on based on stoptimes (realtime or
>    scheduled).
>    4. Check if we are staying on a stop, if we are, return the station
>    stop location.
>    5. If we are in the middle of two stops, find the points closest to
>    our current start and end stops on the geometry line.
>    6. Get the segment where we are on the geometry line.
>    7. Calculate our location on this line.
>
> The algorithm returns an array of objects with the following properties:
>
>    - location: An object with lat and lon as floats.
>    - progress: Fragment of how much of the current segment has been
>    travelled as float.
>    - bearing: The current direction in degrees from north as float.
>    - moving: Whether the vehicle is moving or in station as boolean.
>
> E.g.
> {
> "location: {
> "lat": 60.62200750121682,
> "lon": 24.872947068716414
> },
> "progress": 0.875,
> "bearing": -38.58968855167587,
> "moving": true
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/HSLdevcom/digitransit-ui/issues/461#issuecomment-226348949>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABPQWq0xwIKgOR8RGhvtFXDAB2m6hXoOks5qMIoYgaJpZM4H3zBc>
> .
>


-- 

--
Samuli Heljo
+358 50 3537299

**Comment by tuukka** [comment]
You can do this simulation of vehicle movements outside of Digitransit and use it as an input feed just like real data. For simulation code, see http://14142.net/kartalla/

**Comment by Macroz** [comment]
I think the idea about a implementing a separate "fake-realtime" service is good. It has good separation of concerns. Of course there needs to be some way to tell the real information from the fake information in the UI because of the UX concerns.

As far as the implementation of faking the positions, that's actually not a difficult technical problem. Just interpolate between the points in the routes based on time with a nice formula. But it may be difficult to get good UX out of it. If implemented as a separate service it would be easy to prototype completely separately.

**Comment by kalon33** [comment]
I think that the idea that we could use it as a fallback when Real Time data is not available is important: when RT is not available at all (in the situations I have in France), but also when it is only available for some lines but not others (to have a more uniform UX between transit lines), with the need to clearly indicate that this is not real time data. 

----- Mail original -----

De: "Markku Rontu" <notifications@github.com> 
À: "HSLdevcom/digitransit-ui" <digitransit-ui@noreply.github.com> 
Cc: "Nicolas Derive" <kalon33@ubuntu.com>, "Author" <author@noreply.github.com> 
Envoyé: Jeudi 16 Juin 2016 08:26:39 
Objet: Re: [HSLdevcom/digitransit-ui] Add vehicle position on map (#461) 



I think the idea about a implementing a separate "fake-realtime" service is good. It has good separation of concerns. Of course there needs to be some way to tell the real information from the fake information in the UI because of the UX concerns. 

As far as the implementation of faking the positions, that's actually not a difficult technical problem. Just interpolate between the points in the routes based on time with a nice formula. But it may be difficult to get good UX out of it. If implemented as a separate service it would be easy to prototype completely separately.

**Comment by Macroz** [comment]
One of the UX concerns is that the user cannot trust in the realtime information anymore (because it may be faked). So isn't it just confusing and realtime information should especially reduce stress by showing accurate information. Also at the moment I don't think showing the vehicles all the time on the map is actually very useful at all. Especially if they are mostly faked.

**Comment by kalon33** [comment]
That's why it should be associated to a different color and/or a message indicated it.

**Comment by Macroz** [comment]
Yes, but there's plenty of colors and symbols already creating a pretty high cognitive load :-)
